### PR TITLE
[CELEBORN-583] Merge pooled memory allocators.

### PR DESCRIPTION
**Pull request by FMX**
### What changes were proposed in this pull request?
Merging pooled memory allocators to improve memory usage efficiency. Due to Netty's memory allocator, it will cause some fragments and cannot be used.

### Why are the changes needed?
Workload: 8x Flink TPC-DS 1TB 
Hardware: 8x 32vCPU 128GRam.
Celeborn Worker: 1G off-heap.

Before this patch, as you can see that workers have multiple memory allocators. Memory allocators will consume more memory than we thought it should be. And the workload will failed to run through q4.sql.
截屏2023-05-17 11 06 52
截屏2023-05-15 17 51 30

After this Path, running the same workload will not cause PausePushData state and q4.sql is passed.
截屏2023-05-17 12 28 32

### Does this PR introduce _any_ user-facing change?
NO.

### How was this patch tested?
UT and cluster.

### CODE REVIEW

**Comment by waitinfuture** [comment]
Hi @FMX , thanks for the PR! Could you provide metrics_NettyMemory_Value before this PR?

**Comment by FMX** [comment]
> Hi @FMX , thanks for the PR! Could you provide metrics_NettyMemory_Value before this PR?

Graph updated. I'll provide performance test results later.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 1]
```diff
@@ -796,6 +796,15 @@ class CelebornConf(loadDefaults: Boolean) extends Cloneable with Logging with Se
   def readBufferTargetNotifyThreshold: Long = get(WORKER_READBUFFER_TARGET_NOTIFY_THRESHOLD)
   def readBuffersToTriggerReadMin: Int = get(WORKER_READBUFFERS_TOTRIGGERREAD_MIN)
 
+  def pooledAllocatorArenas: Int = get(MEMORY_POOLED_ALLOCATOR_ARENAS).getOrElse(
+    if (Runtime.getRuntime.availableProcessors() >= 2) {
+      Runtime.getRuntime.availableProcessors()
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:801]: use max to simplify it

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -3204,6 +3213,22 @@ object CelebornConf extends Logging {
       .intConf
       .createWithDefault(32)
 
+  val MEMORY_POOLED_ALLOCATOR_VERBOSE_METRIC: ConfigEntry[Boolean] =
+    buildConf("celeborn.pooled.allocator.verbose.metric")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:3217]: checked `PooledByteBufAllocator`, netty uses `io.netty.allocator.*`, can we use `celeborn .memory.allocator.*` and align the properties w/ netty? e.g. 
`io.netty.allocator.numHeapArenas` => `celeborn .memory.allocator.numHeapArenas`

File: common/src/main/java/org/apache/celeborn/common/network/util/NettyUtils.java [hunk 3]
```diff
@@ -106,4 +110,20 @@ public static PooledByteBufAllocator createPooledByteBufAllocator(
         allowCache ? PooledByteBufAllocator.defaultNormalCacheSize() : 0,
         allowCache ? PooledByteBufAllocator.defaultUseCacheForAllThreads() : false);
   }
+
+  public static PooledByteBufAllocator getShardPooledByteBufAllocator(AbstractSource source) {
+    synchronized (PooledByteBufAllocator.class) {
+      if (_allocator == null) {
+        CelebornConf conf = new CelebornConf();
```
Comment (pan3793) [inline common/src/main/java/org/apache/celeborn/common/network/util/NettyUtils.java:117]: any chance to pass a conf instead create a new one?

**Comment by FMX** [comment]
I ran the first 10 queries of TPC-DS for comparison.
Results from main branch:
main

Results from branch 583:


The total time is quiet close so I think there is no performance degradation.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 2 above)
Comment (FMX) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:3217]: Sounds reasonable. I'll change it to make it looks consistent with Netty.

File: common/src/main/java/org/apache/celeborn/common/network/util/NettyUtils.java [hunk 4]
```diff
@@ -106,4 +110,18 @@ public static PooledByteBufAllocator createPooledByteBufAllocator(
...
   }
+
+  public static PooledByteBufAllocator getShardPooledByteBufAllocator(
+      CelebornConf conf, AbstractSource source) {
+    synchronized (PooledByteBufAllocator.class) {
+      if (_allocator == null) {
+        // each core should have one arena to allocate memory
+        _allocator = createPooledByteBufAllocator(true, true, conf.allocatorArenas());
+        new NettyMemoryMetrics(_allocator, "common-pool", conf.allocatorVerboseMetric(), source);
+        return _allocator;
```
Comment (waitinfuture) [inline common/src/main/java/org/apache/celeborn/common/network/util/NettyUtils.java:121]: We can remove the else branch by moving ```return _allocator;``` out of the if block. IDE says.

**Review by waitinfuture — APPROVED** [review]
LGTM except a nit, thanks!

**Review by waitinfuture — APPROVED** [review]
LGTM except a nit, thanks!

File: common/src/main/java/org/apache/celeborn/common/network/util/NettyUtils.java (hunk 3 above)
Comment (waitinfuture) [reply common/src/main/java/org/apache/celeborn/common/network/util/NettyUtils.java]: +1

**Comment by waitinfuture** [comment]
Shall we add a config to decide whether to use shared allocator? I'm not sure for worker with hundreds of netty threads, a single allocator will be a bottleneck.

**Comment by FMX** [comment]
> Shall we add a config to decide whether to use shared allocator? I'm not sure for worker with hundreds of netty threads, a single allocator will be a bottleneck.

Sounds reasonable. I'll add a new config.

**Comment by waitinfuture** [comment]
LGTM, thanks! Merging to main.

**Pull request merged**